docs/virtual-package.md

```diff
@@ -0,0 +1,9 @@
+# Composer Virtual Packages
+
+Virtual packages are a way to specify the dependency on an implementation of an interface-only repository without forcing a specific implementation. For Httplug, the virtual package is called `php-http/client-implementation`. 
+
+There is no project registered with that name. However, all client implementations including client adapters for Httplug use the `provide` section to tell composer that they do provide the client-implementation.
+
+When *writing a reusable library*, the `require` section should only mention `php-http/client-implementation`, while the `require-dev` section needs to specify a concrete implementation in most cases, in order for the package to be installable on its own for development and testing.
+
+When *using a reusable library* in an application, the `require` section needs to include a concrete implementation in order for the application to install. Missing to do that will lead to composer reporting a missing package.
```

Comment:
i think the error message is a very good idea for one reason: people will google that error message and they should find a page that explains what is going wrong. would be awesome if one day we could register virtual packages in packagist and have composer output an explanation, but until that day we should have this